packages/eslint-plugin/src/rules/no-confusing-void-expression.ts

```diff
@@ -376,5 +422,44 @@ export default createRule<Options, MessageId>({
       const type = getConstrainedTypeAtLocation(services, targetNode);
       return tsutils.isTypeFlagSet(type, ts.TypeFlags.VoidLike);
     }
+
+    function hasValidReturnType(
+      node:
+        | TSESTree.FunctionExpression
+        | TSESTree.ArrowFunctionExpression
+        | TSESTree.FunctionDeclaration
+        | null,
+    ): boolean {
+      const services = getParserServices(context);
+
+      if (node != null) {
+        const functionTSNode = services.esTreeNodeToTSNodeMap.get(node);
+        const functionType = services.getTypeAtLocation(node);
+
+        if (functionTSNode.type) {
+          const signatures = tsutils.getCallSignaturesOfType(functionType);
+
+          return !signatures.every(signature =>
+            tsutils.isTypeFlagSet(
+              signature.getReturnType(),
+              ts.TypeFlags.Any | ts.TypeFlags.Unknown,
+            ),
+          );
+        }
+
+        if (
+          ((node.type === AST_NODE_TYPES.FunctionExpression ||
+            node.type === AST_NODE_TYPES.ArrowFunctionExpression) &&
+            isValidFunctionExpressionReturnType(node, {
+              allowTypedFunctionExpressions: true,
+            })) ||
+          ancestorHasReturnType(node)
+        ) {
+          return true;
+        }
```

Comment:
If we check that function signature don't contain `any` and `unknown` in the return type, we should do the same for typed functions/variable declarations. The following cases are not reported:

```ts
type HigherOrderType = () => any;
(): HigherOrderType => () => console.log();

const x: HigherOrderType = () => console.log();

```
[playground](https://deploy-preview-8632--typescript-eslint.netlify.app/play/#ts=5.4.3&fileType=.ts&code=C4TwDgpgBAEglgcwBYQE4HlUBM0BVzQC8UAFAJRSEB8UAhgHYgDcAUOQFyyIobZ4GUa5QVADGAe3oBncQBsIAOlniE5ViwnTgUAB6d4yNJhyp8kSqQrUxkmfKUq1QA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1rI6YDNYyZgHNaANw6UAJvQAexaCiG90YANrhsORNGgdokADSatWLdkiVhnRQDVJUgJJN701fjiIT2AL7eAupp%2BPkA&tsconfig=&tokens=false)